### PR TITLE
feat(tile-converter): add convertion status dump file

### PR DESCRIPTION
**Pull request by maxkuznetsov-actionengine**

### CODE REVIEW

File: modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
+import {isMap} from 'util/types';
```
Comment (belom88) [inline modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts:1]: Add test

**Review by belom88 — CHANGES_REQUESTED** [review]

File: modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+/**
+ * Convert dump map to object for stringifying to JSON
+ * @param tilesConverted - Map of tilesConverted with conversion status
+ * @returns tilesConverted Map converted to Object
+ */
+export const dumpToObject = (tilesConverted: Map<string, any>) => {
+  const object = {};
+  for (const [key, value] of tilesConverted) {
+    if (value && value.nodes) {
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts:11]: Done

File: modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import {isMap} from 'util/types';
+
+/**
+ * Convert dump map to object for stringifying to JSON
+ * @param tilesConverted - Map of tilesConverted with conversion status
+ * @returns tilesConverted Map converted to Object
+ */
+export const dumpToObject = (tilesConverted: Map<string, any>) => {
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts:8]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 4]
```diff
@@ -756,7 +792,7 @@ export default class I3SConverter {
       nodes.push(node);
 
       if (nodeInPage.mesh) {
-        await this._writeResources(resources, node.id);
+        this._writeResources(resources, node.id, sourceTile);
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts:795]: Reverted back

File: modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts (hunk 1 above)
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts:1]: Done

**Review by belom88 — CHANGES_REQUESTED** [review]

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 5]
```diff
@@ -258,6 +265,17 @@ export default class I3SConverter {
       this.nodePages.useWriteFunction(writeFileForSlpk);
     }
 
+    //create a dump file with convertion options
+    try {
+      writeFile(
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts:270]: Await is already there inside writeFile

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 6]
```diff
@@ -1018,7 +1142,20 @@ export default class I3SConverter {
         case 'jpg':
         case 'png': {
           formats.push({name: '0', format});
-          await this.writeTextureFile(textureData, '0', format, childPath, slpkChildPath);
+          for (const node of dumpTileRecord.nodes) {
+            if (node.nodeId === nodeId) {
+              node.done.set(`${ResourceType.TEXTURE}/${format}`, false);
+            }
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts:1148]: Done

File: modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
+import {isMap} from 'util/types';
+
+/**
+ * Convert dump map to object for stringifying to JSON
+ * @param tilesConverted - Map of tilesConverted with conversion status
+ * @returns tilesConverted Map converted to Object
+ */
+export const dumpTilesToObject = (tilesConverted: Map<string, any>) => {
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/helpers/dump-parser.ts:8]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 8]
```diff
@@ -165,6 +168,7 @@ export default class I3SConverter {
     this.generateBoundingVolumes = false;
     this.layersHasTexture = false;
     this.compressList = null;
+    this.tilesConverted = new Map();
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts:171]: Done

File: modules/tile-converter/src/lib/utils/write-queue.ts [hunk 9]
```diff
@@ -6,6 +7,7 @@ const MEMORY_LIMIT = 4 * 1024 * 1024 * 1024; // 4GB
 
 export type WriteQueueItem = {
   archiveKey?: string;
+  convertedTileDumpMap?: any;
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/lib/utils/write-queue.ts:10]: Done

File: modules/tile-converter/src/lib/utils/write-queue.ts [hunk 10]
```diff
@@ -107,4 +113,34 @@ export default class WriteQueue<T extends WriteQueueItem> extends Queue<T> {
       }
     }
   }
+
+  private updateConvertedTilesMap(
+    convertedTileDumpMaps: any[],
+    writeResults: PromiseSettledResult<string | null>[]
+  ) {
+    for (let i = 0; i < convertedTileDumpMaps.length; i++) {
+      if (convertedTileDumpMaps[i] && 'value' in writeResults[i]) {
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/lib/utils/write-queue.ts:122]: I've move the whole method to the ConversionDump class

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 11]
```diff
@@ -930,37 +992,74 @@ export default class I3SConverter {
...
+   * @param dumpTileRecord - a record of the tilesConverted Map
+   * @param nodeId - nodeId of a converted node for the writing
    */
   private async _writeGeometries(
     geometryBuffer: ArrayBuffer,
     compressedGeometry: Promise<ArrayBuffer>,
     childPath: string,
-    slpkChildPath: string
+    slpkChildPath: string,
+    dumpTileRecord: any,
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts:1003]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 12]
```diff
@@ -918,10 +945,30 @@ export default class I3SConverter {
...
+      slpkChildPath,
+      sourceTile.id!,
+      parseInt(nodePath)
+    );
+    await this._writeShared(
+      sharedResources,
+      childPath,
+      slpkChildPath,
+      nodePath,
+      sourceTile.id!,
```
Comment (belom88) [inline modules/tile-converter/src/i3s-converter/i3s-converter.ts:961]: "!" is a potential bug. If the `id` is nullable we'd better use url/filepath of the input file

**Review by belom88 — CHANGES_REQUESTED** [review]

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 13]
```diff
@@ -756,7 +766,18 @@ export default class I3SConverter {
       nodes.push(node);
 
       if (nodeInPage.mesh) {
-        await this._writeResources(resources, node.id);
+        //update a record in a dump file
+        const nodeId = parseInt(node.id);
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts:770]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 14]
```diff
@@ -756,7 +766,18 @@ export default class I3SConverter {
...
+        if (!isNaN(nodeId) && sourceTile.id) {
+          const {nodes} = this.conversionDump.getRecord(sourceTile.id) || {nodes: []};
+          nodes.push({nodeId, done: {}});
+          if (nodes.length === 1) {
+            this.conversionDump.setRecord(sourceTile.id, {nodes});
+          }
+          this.conversionDump.updateDumpFile();
+        }
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts:778]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts [hunk 15]
```diff
@@ -1247,6 +1407,7 @@ export default class I3SConverter {
     console.log(`File(s) size: `, filesSize, ' bytes'); // eslint-disable-line no-undef, no-console
     console.log(`Percentage of tiles with "ADD" refinement type:`, addRefinementPercentage, '%'); // eslint-disable-line no-undef, no-console
     console.log(`------------------------------------------------`); // eslint-disable-line no-undef, no-console
+    this.conversionDump.deleteDumpFile();
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts:1410]: Done

File: modules/tile-converter/src/lib/utils/conversion-dump.ts [hunk 16]
```diff
@@ -0,0 +1,182 @@
+import {DUMP_FILE_SUFFIX} from '../../constants';
```
[the remark this answers is not in the document]
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/lib/utils/conversion-dump.ts:1]: Done

File: modules/tile-converter/src/i3s-converter/i3s-converter.ts (hunk 12 above)
Comment (maxkuznetsov-actionengine) [reply modules/tile-converter/src/i3s-converter/i3s-converter.ts]: Done

**Review by belom88 — APPROVED** [review]

**Pull request merged**